db/compaction/sst_partitioner.cc

```diff
@@ -0,0 +1,70 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+
+#include "rocksdb/sst_partitioner.h"
+
+#include <algorithm>
+
+namespace ROCKSDB_NAMESPACE {
+
+class SstPartitionerFixedPrefix : public SstPartitioner {
+ public:
+  SstPartitionerFixedPrefix(size_t len) : len_(len) {}
+
+  virtual ~SstPartitionerFixedPrefix(){};
+
+  const char* Name() const override { return "SstPartitionerFixedPrefix"; }
+
+  PartitionerResult ShouldPartition(
+      const PartitionerRequest& request) override {
+    Slice last_key_fixed(*request.prev_user_key);
+    if (last_key_fixed.size() > len_) {
+      last_key_fixed.size_ = len_;
+    }
+    Slice current_key_fixed(*request.current_user_key);
+    if (current_key_fixed.size() > len_) {
+      current_key_fixed.size_ = len_;
+    }
+    return last_key_fixed.compare(current_key_fixed) != 0 ? kRequired
+                                                          : kNotRequired;
+  }
+
+  bool CanDoTrivialMove(const Slice& smallest_user_key,
+                        const Slice& largest_user_key) override {
+    return ShouldPartition(PartitionerRequest(
+               smallest_user_key, largest_user_key, 0)) == kNotRequired;
+  }
+
+ private:
+  size_t len_;
+};
+
+class SstPartitionerFixedPrefixFactory : public SstPartitionerFactory {
```

Comment:
"Struct and class definitions should either be defined in a header file, declared in a header file with the same base name and path as the definition file, or in an anonymous namespace to prevent possible shadowing of other classes."

Reply:
Fixed